### PR TITLE
Remove legacy meta data from orders

### LINKED ISSUE #20119: privacy data eraser for orders doesn't remove legacy paypal metadata
## Summary
Woo's new privacy data eraser for orders doesn't remove legacy Paypal gateway metadata.
In Woo's past, four items of PII from Paypal were stored in the order metadata under the following custom fields

```
Payer first name
Payer last name
Payer PayPal address
Transaction ID
```

A fifth, `Payment type` was also stored but is not PII. These were never migrated/purged in newer Woo versions. Therefore, this data continues to exist in these legacy fields on older orders paid via the Paypal gateway.

See code from Woo v2.0.6 -> v2.2.3 at https://github.com/woocommerce/woocommerce/commit/dfab71586b369d4568cf489f40347a41d10a274b#diff-93922b66b3e06cd0aad573d1ff782985

And the current privacy eraser code at https://github.com/woocommerce/woocommerce/blob/fdb8cdb482c61db4fc464dedb9736c78871d83f7/includes/class-wc-privacy-erasers.php#L255

## Setup

* Woo v3.3.5 store that has be upgraded over time from the v2 days and always used the Paypal gateway
* Wordpress 4.9.5

## Repro

1. Open an order that is several years old and made in the v2 days with Paypal
2. View the custom fields

## Result

The four PII sensitive fields are visible on the order and will not be deleted by the privacy eraser

## Expected

The four PII sensitive fields to be erased.

## Isolating the problem (mark completed items with an [x]):
- [X] I can reproduce this bug consistently using the steps above.

## Resolutions

I recommend one of the following:
1. Migrate this legacy Paypal data into new fields which the eraser will manage
2. Add code to anon/remove this legacy metadata into the order eraser. I don't think the fields can be added into a custom filter for `woocommerce_privacy_remove_order_personal_data_props` since these are not standard fields which have `order->get_xxxx` methods.
3. Add documentation stating this issue and suggesting the customer write a custom action handler for `woocommerce_privacy_before_remove_order_personal_data` or `woocommerce_privacy_remove_order_personal_data` to anon/remove these legacy metadata fields

## WordPress Environment

<details>

### WordPress Environment ###

Home URL: http://ubuntu1604
Site URL: http://ubuntu1604
WC Version: 3.3.5
Log Directory Writable: ✔
WP Version: 4.9.5
WP Multisite: –
WP Memory Limit: 128 MB
WP Debug Mode: ✔
WP Cron: –
Language: en_US

### Server Environment ###

Server Info: Apache/2.4.18 (Ubuntu)
PHP Version: 7.0.30-0ubuntu0.16.04.1
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.3.5
WC Database Prefix: wp_
Total Database Size: 18.78MB
Database Data Size: 9.87MB
Database Index Size: 8.91MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.14MB + Index: 0.23MB
wp_woocommerce_order_items: Data: 0.17MB + Index: 0.08MB
wp_woocommerce_order_itemmeta: Data: 1.52MB + Index: 2.00MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.27MB + Index: 0.30MB
wp_comments: Data: 1.52MB + Index: 0.66MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 0.28MB + Index: 0.05MB
wp_postmeta: Data: 2.52MB + Index: 4.03MB
wp_posts: Data: 1.45MB + Index: 0.25MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.06MB + Index: 0.05MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 1.52MB + Index: 0.77MB
wp_users: Data: 0.06MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_wpmm_subscribers: Data: 0.02MB + Index: 0.00MB

### Post Type Counts ###

attachment: 63
cookielawinfo: 6
custom_css: 2
nav_menu_item: 23
page: 23
post: 10
product: 8
product_variation: 5
revision: 138
shop_coupon: 4
shop_order: 745
shop_order_refund: 1

### Security ###

Secure connection (HTTPS): ❌Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (14) ###

Akismet Anti-Spam: by Automattic – 4.0.3
All In One SEO Pack: by Michael Torbert – 2.5
Change Wordpress robot email behavior: by Dale Phurrough – 1.0.1
Cookie Law Info: by Richard Ashby – 1.5.3
Disable REST API: by Dave McHale – 1.4.3
Disable XML-RPC: by Philip Erb – 1.0.1
Hard 404 on Wordpress searches: by Dale Phurrough – 1.00
Reset Opcache on Upgrade: by Dale Phurrough – 1.1.0
Google Analytics for WordPress by MonsterInsights: by MonsterInsights – 7.0.6
Woocommerce layout changes for hidale.com: by Dale Phurrough – 1.0.2 – Not tested with the active version of WooCommerce
Woocommerce Paypal Gateway Extras: by Dale Phurrough – 1.1.0 – Not tested with the active version of WooCommerce
Woocommerce Digital Downloads, Registration, Authorization: by Dale Phurrough – 1.1.0 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.3.5
WP Google Authenticator: by Julien Liabeuf – 1.1.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop base: #753 - /shop/
Cart: #754 - /cart/
Checkout: #755 - /checkout/
My account: #756 - /my-account/
Terms and conditions: #750 - /terms/

### Theme ###

Name: Twenty Eleven Customized by Dale
Version: 1.0
Author URL: http://hidale.com
Child Theme: ✔
Parent Theme Name: Twenty Eleven
Parent Theme Version: 2.7
Parent Theme Author URL: https://wordpress.org/
WooCommerce Support: ✔

### Templates ###

Overrides: –

</details>

### PR DESCRIPTION
**Pull request by mikejolley**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

I've included the types for future proofing but this just deletes meta :)

Closes #20119

### How to test the changes in this Pull Request:

1. Add custom field to an order called Payer last name
2. Remove order data via the bulk actions on order screen
3. Check data is removed.

### CODE REVIEW

**Comment by kloon** [comment]
Should we not rather make the gateway use woocommerce_privacy_remove_order_personal_data_meta as some of the data in here are specific to that gateway? Transaction ID is fine since that is a core field.

**Comment by diablodale** [comment]
@kloon, I agree with the intention - that the owner/maker is responsible for cleanup. If moved to the paypal gateway, this erase code would need to still register/run even if the gateway is not enabled. It is possible that in the Woo v2.x days a store used the paypal gateway, yet this month it doesn't.

File: includes/class-wc-privacy-erasers.php [hunk 1]
```diff
@@ -295,12 +294,35 @@ public static function remove_order_personal_data( $order ) {
...
+				 * @param bool     $anonymized_data Value of this data after anonymization.
+				 * @param string   $prop Meta key being removed.
+				 * @param string   $value Current value of the data.
+				 * @param string   $data_type Type of data.
+				 * @param WC_Order $order An order object.
+				 */
+				$anon_value = apply_filters( 'woocommerce_privacy_remove_order_personal_data_meta_value', $anon_value, $meta_key, $value, $data_type, $order );
+
+				if ( $anon_value ) {
+					$order->delete_meta_data( $meta_key );
```
Comment (diablodale) [inline includes/class-wc-privacy-erasers.php]: lines 319 and 321 appear reversed. If there is an anon_value, it should be updated, else deleted.

File: includes/class-wc-privacy-erasers.php [hunk 2]
```diff
@@ -287,6 +283,46 @@ public static function remove_order_personal_data( $order ) {
...
+					continue;
+				}
+
+				$anon_value = function_exists( 'wp_privacy_anonymize_data' ) ? wp_privacy_anonymize_data( $data_type, $value ) : '';
+
+				/**
+				 * Expose a way to control the anonymized value of a value via 3rd party code.
+				 *
+				 * @since 3.4.0
+				 * @param bool     $anonymized_data Value of this data after anonymization.
```
Comment (kloon) [inline includes/class-wc-privacy-erasers.php]: This should be a string no? Also the variable name does not match the actual variable being used.

File: includes/class-wc-privacy-erasers.php [hunk 3]
```diff
@@ -287,6 +283,46 @@ public static function remove_order_personal_data( $order ) {
...
+				 * @param string   $prop Meta key being removed.
```
Comment (kloon) [inline includes/class-wc-privacy-erasers.php]: Variable name does not match actual variable being used.

File: includes/class-wc-privacy-erasers.php [hunk 4]
```diff
@@ -287,6 +283,46 @@ public static function remove_order_personal_data( $order ) {
...
+				 * @param string   $value Current value of the data.
+				 * @param string   $data_type Type of data.
+				 * @param WC_Order $order An order object.
+				 */
+				$anon_value = apply_filters( 'woocommerce_privacy_remove_order_personal_data_meta_value', $anon_value, $meta_key, $value, $data_type, $order );
+
+				if ( $anon_value ) {
```
Comment (kloon) [inline includes/class-wc-privacy-erasers.php]: This should be ! $anon_value

**Review by kloon — DISMISSED** [review]
Small couple of tweaks needed.

File: includes/class-wc-privacy-erasers.php (hunk 4 above)
Comment (rodrigoprimo) [reply includes/class-wc-privacy-erasers.php]: It seems to me that the logic is correct. If there is an anonymized version of the value, update the original value. If there is not, delete the entry.

**Review by rodrigoprimo — APPROVED** [review]
LGTM

**Pull request merged**